### DISCUSSION TITLE
Conditional SQL param throws IndeterminateDatatypeError

### DISCUSSION THREAD

**Discussion by gregbrowndev** [sqlalchemy/sqlalchemy#9018] [Usage Questions]
Python 3.11.0
SqlAlchemy: v1.4.44

Hi all,

I'm running into a problem trying to execute a parameterised SQL statement. I want the parameter to conditionally filter the resultset, i.e. if the param is `None`, don't filter, and if it's not `None`, do filter.

Let's say you have a table of cats:

```sql
CREATE TABLE cats (
    name varchar,
    size varchar
);

INSERT INTO cats VALUES ('Mac', 'large'), ('Podge', 'x-large'), ('Tidge', 'small');
```

This code works fine when the `size` param is always required:

```py
statement = sa.text("SELECT * FROM cats WHERE size = :size")
result = session.execute(statement, {"size": "large"})
```

However, if I make the param conditional:

```py
statement = sa.text("SELECT * FROM cats WHERE (:size is NULL or size = :size)")
result = session.execute(statement, {"size": "large"})
```

then I get this error:

```
sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.IndeterminateDatatypeError'>: could not determine data type of parameter $1
[SQL: SELECT * FROM cats
WHERE (%s is NULL or size = %s)
]
[parameters: ('large', 'large')]
```

I've also tried explicitly setting the type of the param using [bindparams](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.TextClause.bindparams). However, that didn't fix the issue.

Appreciate any help thanks!

**Comment by lelit** [comment]
That seems a problem with the underlying `asyncpg` driver: I cannot verify right now, but I use something similar with `psycopg2` (thru bindparams) and never had that issue. Can you try executing the same query directly on the `asyncpg` connection?

**Comment by CaselIT** [comment]
Hi,

The error comes from asyncpg/postgresql. Try adding a cast in the text sql
